### PR TITLE
fix(fabric): forward set_epoch to underlying sampler in DistributedSamplerWrapper

### LINKED ISSUE #21454: DistributedSamplerWrapper does not pass on the `.set_epoch` call to the underlying sampler
### Bug description

By default the Trainer calls `.set_epoch` of any simple sampler [in the fit_loop](https://github.com/Lightning-AI/pytorch-lightning/blob/027455bcd9433f201b1136f68d54b1a07588abe3/src/lightning/pytorch/loops/fit_loop.py#L441).
However when a sampler is wrapped by [DistributedSamplerWrapper](https://github.com/Lightning-AI/pytorch-lightning/blob/027455bcd9433f201b1136f68d54b1a07588abe3/src/lightning/fabric/utilities/distributed.py#L354) it ignores passing this call to the underlying sampler. I suggest the following fix for this purpose:

```
class DistributedSamplerWrapper(DistributedSampler):
    # current implementation

    @override
    def set_epoch(self, epoch: int) -> None:
        super().set_epoch(epoch)
        self.dataset._sampler.set_epoch(epoch)
```

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

```
# Error messages and logs here please
```

### Environment

<details>
  <summary>Current environment</summary>

```
#- PyTorch Lightning Version (e.g., 2.5.0):
#- PyTorch Version (e.g., 2.5):
#- Python version (e.g., 3.12):
#- OS (e.g., Linux):
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
```

</details>

### More info

_No response_

cc @ethanwharris @justusschock

### PR DESCRIPTION
**Pull request by littlebullGit**
What does this PR do?

The `DistributedSamplerWrapper` now forwards `set_epoch()` calls to the original sampler if it supports the method.

### Problem

When a custom sampler is wrapped by `DistributedSamplerWrapper`, calling `set_epoch()` on the wrapper does not propagate to the original sampler. This breaks samplers that rely on `set_epoch()` for epoch-dependent behavior like shuffling.

### Solution

Override `set_epoch()` in `DistributedSamplerWrapper` to:

1. Call `super().set_epoch(epoch)` to set the epoch on the wrapper itself
2. Forward the call to the original sampler if it has a callable `set_epoch` method

This fix is **generic** and works for any sampler subclass that implements `set_epoch()`, not just specific implementations. It uses duck typing to check for the method's existence and callability.

### Changes

- **`src/lightning/fabric/utilities/distributed.py`**: Added `set_epoch()` override to `DistributedSamplerWrapper`
- **`src/lightning/fabric/CHANGELOG.md`**: Added changelog entry
- **`tests/tests_fabric/utilities/test_distributed.py`**: Added `test_distributed_sampler_wrapper_set_epoch` with 100% branch coverage:
  - Case 1: Sampler WITH `set_epoch` method - verifies forwarding works
  - Case 2: Sampler WITHOUT `set_epoch` method - verifies graceful handling
  - Case 3: Sampler with non-callable `set_epoch` attribute - verifies it's skipped

Fixes #21454

### CODE REVIEW

File: src/lightning/fabric/utilities/distributed.py [hunk 1]
```diff
@@ -372,6 +372,14 @@ def __iter__(self) -> Iterator:
         self.dataset.reset()
         return (self.dataset[index] for index in super().__iter__())
 
+    @override
+    def set_epoch(self, epoch: int) -> None:
+        super().set_epoch(epoch)
+        # Forward set_epoch to the underlying sampler if it supports it
+        underlying_sampler = self.dataset._sampler
```
Comment (deependujha) [inline src/lightning/fabric/utilities/distributed.py]: nitpick, should we use `original_sampler` or something similar than `underlying_sampler`, lol!

File: tests/tests_fabric/utilities/test_distributed.py [hunk 2]
```diff
@@ -274,3 +275,117 @@ def test_is_dtensor(monkeypatch):
...
+    # A simple sampler without set_epoch method
+    class SimpleSampler(torch.utils.data.Sampler):
+        def __init__(self, data_source):
+            self.data_source = data_source
+
+        def __len__(self):
+            return len(self.data_source)
+
+        def __iter__(self):
+            return iter(range(len(self.data_source)))
```
Comment (deependujha) [inline tests/tests_fabric/utilities/test_distributed.py]: I believe we can refactor tests to be more compact and still cover all the cases.

Like, having a `CustomSampler` and subclass `CustomSamplerWithSetEpoch`.

```python
class CustomSampler(torch.utils.data.Sampler):
        def __init__(self, data_source, use_set_epoch: bool = False):
            self.data_source = data_source
            if use_set_epoch:
                  self.set_epoch = "not a method"  # attribute exists but is not callable
            ...

class CustomSamplerWithSetEpoch(CustomSampler):
        def set_epoch(self, epoch):
            self.epoch = epoch
            self.set_epoch_call_count += 1
```

And almost all the tests can be combined and assert behaviour in a single test with `pytest parameterize`.

What do you think? :)

File: src/lightning/fabric/utilities/distributed.py (hunk 1 above)
Comment (littlebullGit) [reply src/lightning/fabric/utilities/distributed.py]: sure. done

**Comment by deependujha** [comment]
Please update changelog too. :)

File: tests/tests_fabric/utilities/test_distributed.py (hunk 2 above)
Comment (littlebullGit) [reply tests/tests_fabric/utilities/test_distributed.py]: Merged 3 separate tests into 1 test with 3 cases:
  - Case 1: Sampler WITH `set_epoch` method
  - Case 2: Sampler WITHOUT `set_epoch` method  
  - Case 3: Sampler with non-callable `set_epoch` attribute

**Comment by littlebullGit** [comment]
> Please update changelog too. :)

changelog is the main source of the merge conflict. So I am kind of hesitate to update it until someone reviewed. Not sure what is the best way to handle it. Updated.

**Review by deependujha — APPROVED** [review]

**Comment by deependujha** [comment]
Yeah, but without it, we sometimes merge the PR without updates in changelog.

**Comment by littlebullGit** [comment]
> Yeah, but without it, we sometimes merge the PR without updates in changelog.
Maybe we should add the changelog entry in the PR description and whoever merge it take it and add the the changelog ? just a thought.

**Comment by deependujha** [comment]
there're multiple maintainers maintaining this and other Lightning AI's OSS projects, I doubt if one will even remember, lol.

File: src/lightning/fabric/utilities/distributed.py [hunk 3]
```diff
@@ -372,6 +372,14 @@ def __iter__(self) -> Iterator:
...
         return (self.dataset[index] for index in super().__iter__())
 
+    @override
+    def set_epoch(self, epoch: int) -> None:
+        super().set_epoch(epoch)
+        # Forward set_epoch to the original sampler if it supports it
+        original_sampler = self.dataset._sampler
+        if hasattr(original_sampler, "set_epoch") and callable(original_sampler.set_epoch):
+            original_sampler.set_epoch(epoch)
+
```
Comment (bhimrazy) [inline src/lightning/fabric/utilities/distributed.py:382]: I think it would be nice to move this part  to **`_DatasetSamplerWrapper`** and simply call `set_epoch` here.

That said, let’s wait for feedback from the lead maintainers before proceeding with this change.

Comment (deependujha) [reply src/lightning/fabric/utilities/distributed.py:382]: Hi @bhimrazy, but `_DatasetSamplerWrapper` won't be called then.

Calling `set_epoch` on `DistributedSamplerWrapper` and not implementing it, will just use parent (DistributedSampler) `set_epoch` method. But, `_DatasetSamplerWrapper` won't be called.

Let me know if I'm missing something.

Comment (bhimrazy) [reply src/lightning/fabric/utilities/distributed.py:382]: I actually only meant for lines 379–381, as it uses the sampler from `_DatasetSamplerWrapper` itself.
And then replace it with something like `self.dataset.set_epoch(epoch)`.

Comment (deependujha) [reply src/lightning/fabric/utilities/distributed.py:382]: thanks for clarifying. It makes sense, but in that case too, we need to check:

```python
class _DatasetSamplerWrapper:
    ....
    def set_epoch(self, epoch: int):
        if hasattr(self._sampler, "set_epoch") and callable(self._sampler, "set_epoch):
            self._sampler.set_epoch(epoch)
```

Not sure what we would gain, but it makes sense that sampler wrapper to have `set_epoch` method.
